treasury.go

```diff
@@ -0,0 +1,204 @@
+		if !vspPreferenceUpdateSuccess {
+			// Updating the treaury spend voting preference with the vsp failed,
+			// revert the locally saved voting preference for the treasury spend.
+			revertError := wallet.Internal().SetTreasuryKeyPolicy(ctx, pikey, currentVotingPolicy, ticketHash)
+			if revertError != nil {
+				log.Errorf("unable to revert locally saved voting preference: %v", revertError)
+			}
+		}
+	}()
+
+	// If a ticket hash is provided, set the specified vote policy with
+	// the VSP associated with the provided ticket. Otherwise, set the
+	// vote policy with the VSPs associated with all "votable" tickets.
+	ticketHashes := make([]*chainhash.Hash, 0)
+	if ticketHash != nil {
+		ticketHashes = append(ticketHashes, ticketHash)
+	} else {
+		err = wallet.Internal().ForUnspentUnexpiredTickets(ctx, func(hash *chainhash.Hash) error {
+			ticketHashes = append(ticketHashes, hash)
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("unable to fetch hashes for all unspent, unexpired tickets: %v", err)
+		}
+	}
+
+	// Never return errors from this for loop, so all tickets are tried.
+	// The first error will be returned to the caller.
+	var firstErr error
+	for _, tHash := range ticketHashes {
+		vspTicketInfo, err := wallet.Internal().VSPTicketInfo(ctx, tHash)
+		if err != nil {
+			// Ignore NotExist error, just means the ticket is not
+			// registered with a VSP, nothing more to do here.
+			if firstErr == nil && !errors.Is(err, errors.NotExist) {
+				firstErr = err
+			}
+			continue // try next tHash
+		}
+
+		// Update the vote policy for the ticket with the associated VSP.
+		vspClient, err := wallet.VSPClient(vspTicketInfo.Host, vspTicketInfo.PubKey)
+		if err != nil && firstErr == nil {
+			firstErr = err
+			continue // try next tHash
+		}
+		err = vspClient.SetVoteChoice(ctx, tHash, nil, nil, policyMap)
+		if err != nil && firstErr == nil {
+			firstErr = err
+			continue // try next tHash
+		}
+	}
+
+	vspPreferenceUpdateSuccess = firstErr == nil
+	return firstErr
+}
+
+// AllTreasuryPolicies returns voting policies for treasury spends by a particular
+// key. If a key is specified, that policy is returned; otherwise the policies
+// for all keys are returned in an array. If both a key and ticket hash are
+// provided, the per-ticket key policy is returned.
+func (wallet *Wallet) AllTreasuryPolicies(key, hash string) ([]*TreasuryKeyPolicy, error) {
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		var err error
+		ticketHash, err = chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+	}
+
+	if key != "" {
+		pikey, err := hex.DecodeString(key)
+		if err != nil {
+			return nil, fmt.Errorf("parameter contains invalid hexadecimal: %w", err)
+		}
+		var policy string
+		switch wallet.Internal().TreasuryKeyPolicy(pikey, ticketHash) {
+		case stake.TreasuryVoteYes:
+			policy = "yes"
+		case stake.TreasuryVoteNo:
+			policy = "no"
+		default:
+			policy = "abstain"
```

Comment:
default is invalid, or maybe just empty string. "abstain" is not a valid treasury vote type. See `stake.CheckTreasuryVote`.

Reply:
this was implemented upstream https://github.com/decred/dcrwallet/blob/master/internal/rpc/jsonrpc/methods.go#L3760

Reply:
i can't seem to find any reference to `stake.CheckTreasuryVote` in dcrwallet, can you link me